File: leetcode/189/c++/solution.cpp
/*
    189. Rotate Array
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    void rotate(vector<int>& nums, int k) {
        k = k % nums.size();
        int previous = nums[0], temp;
        for (size_t i = 0, current = 0, next = 0, start = 0; i < nums.size();
             ++i) {
            next = (current + k) % nums.size();
            temp = nums[next];
            nums[next] = previous;
            previous = temp;
            current = next;
            if (start == current && i + 1 < nums.size()) {
                ++current;
                ++start;
                previous = nums[start];
            }
        }
    }
};

File: leetcode/189/rust/solution.rs
/*
    189. Rotate Array
    Author: Miguel Angel Bermeo Ayerbe
*/
impl Solution {
    pub fn reverse(nums: &mut Vec<i32>,lb:usize,ub:usize){
        let middle:usize=(ub-lb)/2;
        for i in 0..(middle) {
            nums.swap(i+lb,ub-i-1);
        }
    }
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let k_movs:usize=(k as usize)%nums.len();
        nums.reverse();
        Solution::reverse(nums,0,k_movs);
        Solution::reverse(nums,k_movs,nums.len());
    }

}